dht/byte_ordered_partitioner.hh

```diff
@@ -74,6 +74,7 @@ public:
         }
     }
     virtual unsigned shard_of(const token& t) const override;
+    virtual unsigned shard_of(const token& t, unsigned shard_count, unsigned sharding_ignore_msb) const override;
```

Comment:
Why not create a new partitioner instance configured how you like and use the regular shard_of()?

Reply:
@kbr- ^

Reply:
Because I don't know that if favourite partitioner, the one I would create, is the one that will be used later when token is calculated. So I want to use the partitioner passed from the system (for example, `generate_topology_description` takes `const i_partitioner&`).

But yes, eventually when we introduce custom per-table partitioners and we will know 100% that CDC log tables use one partitioner of our choice, I could go with a particular instance as you said.

Reply:
You could write the three-argument call to shard_of as `byte_ordered_partitioner(shard_count, smp).shard_of(token)`. And it's really a four-argument call, since the type of the partitioner is also a parameter. So you allow overriding two of the partitioner's parameters, but not the third.

Reply:
I don't get your point. Yes, I am given an `i_partitioner&` and I am
overriding only two parameters (shard count and ignore msb), that was my
intent.

I could construct a `byte_ordered_partitioner`, but that's not what I want.
I want to use the *given* partitioner (could be byteordered, could be
murmur3, could be random). Just with these two parameters changed.

Other way would be to perform a bunch of dynamic_casts and based on that,
construct a new partitioner with my two custom parameters. And copy the
rest of parameters from the old instance to the new. Meh

Reply:
No, of course if you are given a partitioner and just those two parameters change, then it makes sense to do what you did. But why is that the case? I guess you are transporting partitioner parameters from other nodes, and the partitioner class can't change but the other parameters can.

Reply:
Yes. I'm using this new `shard_of` method in two places:
1. when choosing streams for a new CDC generation so that each shard on each node has a stream; for this I'm using the node's shard_count and ignore_msb_bits, which I obtain from gossip
2. during CDC log writes, when I choose a stream for the write (`get_stream` in cdc/metadata.cc).

Reply:
See also docs/cdc.md (section "The generation's mapping")